llpc/lower/llpcSpirvLowerConstImmediateStore.cpp

```diff
@@ -51,164 +52,180 @@ namespace Llpc {
 //
 // @param allocaInst : The "alloca" instruction to process
-StoreInst *SpirvLowerConstImmediateStore::findSingleStore(AllocaInst *allocaInst) {
-  std::vector<Instruction *> pointers;
-  bool isOuterPointer = true;
-  StoreInst *storeInstFound = nullptr;
-  Instruction *pointer = allocaInst;
-  while (true) {
-    for (auto useIt = pointer->use_begin(), useItEnd = pointer->use_end(); useIt != useItEnd; ++useIt) {
-      auto user = cast<Instruction>(useIt->getUser());
+// @return true if the alloca was replaced
+bool SpirvLowerConstImmediateStore::tryProcessAlloca(AllocaInst *allocaInst) {
+  auto *allocatedTy = allocaInst->getAllocatedType();
+  auto *arrayTy = dyn_cast<ArrayType>(allocatedTy);
+
+  StoreInst *aggregateStore = nullptr;
+  DenseMap<uint64_t, StoreInst *> elementStores;
+
+  SmallVector<Instruction *> toErase;
+  SmallVector<GetElementPtrInst *> geps;
+
+  // Step 1: Determine if the alloca can be converted and find the relevant store(s)
+  SmallVector<std::pair<Value *, std::optional<uint64_t>>> pointers;
+  pointers.emplace_back(allocaInst, 0);
+  do {
+    auto [pointer, index] = pointers.pop_back_val();
+    for (Use &use : pointer->uses()) {
+      auto user = cast<Instruction>(use.getUser());
       if (auto storeInst = dyn_cast<StoreInst>(user)) {
-        if (pointer == storeInst->getValueOperand() || storeInstFound || !isOuterPointer) {
-          // Pointer escapes by being stored, or we have already found a "store"
-          // instruction, or this is a partial "store" instruction.
-          return nullptr;
+        if (use.getOperandNo() != storeInst->getPointerOperandIndex() || !index.has_value()) {
+          // Pointer escapes by being stored, or we store to a dynamically indexed (or otherwise complex) pointer.
+          return false;
         }
-        storeInstFound = storeInst;
-      } else if (auto getElemPtrInst = dyn_cast<GetElementPtrInst>(user))
-        pointers.push_back(getElemPtrInst);
-      else if (!isa<LoadInst>(user) && !isAssumeLikeIntrinsic(user)) {
-        // Pointer escapes by being used in some way other than "load/store/getelementptr".
-        return nullptr;
-      }
-    }
 
-    if (pointers.empty())
-      break;
+        Value *storeValue = storeInst->getValueOperand();
+        if (!isa<Constant>(storeValue))
+          return false;
 
-    pointer = pointers.back();
-    pointers.pop_back();
-    isOuterPointer = false;
-  }
+        if (storeValue->getType() == allocatedTy) {
+          if (aggregateStore || !elementStores.empty() || index.value() != 0)
+            return false;
+          aggregateStore = storeInst;
+          continue;
+        }
 
-  return storeInstFound;
-}
+        if (arrayTy && storeValue->getType() == arrayTy->getArrayElementType()) {
+          if (aggregateStore)
+            return false;
+          if (index.value() >= arrayTy->getArrayNumElements())
+            return false;
```

Comment:
Is this for bad-written shader code? Can we just erase the instruction and continue processing?

Reply:
Yes, we can do that as well: If the `store` ever executes it would be UB. But perhaps we can have cases where such a `store` is inside of control flow such that it is never executed -- could be the result of some metaprogramming in the shader (perhaps the size of the array is controlled by a #define or specialization constant or something...). So it's more general to follow your suggestion.